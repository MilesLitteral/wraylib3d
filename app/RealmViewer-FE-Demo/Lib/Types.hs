module Lib.Types where

type BooksWenv = WidgetEnv  BooksModel BooksEvt
type BooksNode = WidgetNode BooksModel BooksEvt
    
keyListType c v =
    case c of
      XBOX_CONTROLLER     -> xboxKeyList    v
      PS_CONTROLLER       -> psKeyList      v
      TP_CONTROLLER       -> thirdPartyList v 
      GENERIC_CONTROLLER  -> genericKeyList v

keyTextListType c =
  case c of
    XBOX_CONTROLLER     -> xboxKeyTextList
    PS_CONTROLLER       -> psKeyTextList    
    TP_CONTROLLER       -> thirdKeyTextList 
    GENERIC_CONTROLLER  -> thirdKeyTextList

xboxKeyTextList = [Mapping,        
  START,           
  SELECT,          
  L_DPAD,          
  R_DPAD,          
  U_DPAD,          
  D_DPAD,          
  L_STICK,         
  R_STICK,         
  BUTTON_A, 
  BUTTON_B,   
  BUTTON_X,     
  BUTTON_Y,  
  BUTTON_L1, 
  BUTTON_R1,
  BUTTON_L2, 
  BUTTON_R2,
  BUTTON_L3,
  BUTTON_R3
  ]       

psKeyTextList = [
    Mapping,           
    OPTION,          
    SHARE,           
    HOME_BUTTON,     
    TRACKPAD,         --Playstation Only
    PS_BUTTON_A,     
    PS_BUTTON_B,
    PS_BUTTON_X,     
    PS_BUTTON_Y,     
    PS_BUTTON_L1,    
    PS_BUTTON_R1,    
    PS_BUTTON_L2,    
    PS_BUTTON_R2,    
    PS_BUTTON_L3,    
    PS_BUTTON_R3   
    ] 

thirdKeyTextList =  [
    Mapping,                 
    KEYBOARD_KEY,    
    L_MOUSE_KEY,     
    R_MOUSE_KEY,     
    C_MOUSE_KEY,     
    XR_GESTURE    
    ]    

xboxKeyList c = 
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"                   `styleBasic` [height 15,width 15]
    START           -> image "./assets/images/ControllerIcons/Special/Menu_dark.png"              `styleBasic` [height 15,width 15]
    SELECT          -> image "./assets/images/ControllerIcons/Special/View_dark.png"              `styleBasic` [height 15,width 15]
    L_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_left_dark.png"      `styleBasic` [height 15,width 15]
    R_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_right_dark.png"     `styleBasic` [height 15,width 15]
    U_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_up_dark.png"        `styleBasic` [height 15,width 15]
    D_DPAD          -> image "./assets/images/ControllerIcons/Digipad/Digipad_down_dark.png"      `styleBasic` [height 15,width 15]
    L_STICK         -> image "./assets/images/ControllerIcons/Stick/L_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    R_STICK         -> image "./assets/images/ControllerIcons/Stick/R_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_A        -> image "./assets/images/ControllerIcons/Button/A_color_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_B        -> image "./assets/images/ControllerIcons/Button/B_color_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_X        -> image "./assets/images/ControllerIcons/Button/X_color_dark.png"      `styleBasic` [height 15,width 15]--, sizeReqW $ fixedSize 25, sizeReqH $ fixedSize 25]
    BUTTON_Y        -> image "./assets/images/ControllerIcons/Button/Y_color_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_L1       -> image "./assets/images/ControllerIcons/ShoulderButton/LB_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_R1       -> image "./assets/images/ControllerIcons/ShoulderButton/RB_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_L2       -> image "./assets/images/ControllerIcons/ShoulderButton/LT_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_R2       -> image "./assets/images/ControllerIcons/ShoulderButton/RT_dark.png"      `styleBasic` [height 15,width 15]
    BUTTON_L3       -> image "./assets/images/ControllerIcons/ShoulderButton/windows_sm.png"      `styleBasic` [height 15,width 15]
    BUTTON_R3       -> image "./assets/images/ControllerIcons/ShoulderButton/windows_sm.png"      `styleBasic` [height 15,width 15]

psKeyList c = 
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"                   `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    OPTION          -> image "./assets/images/ControllerIcons/PlaystationButtons/Share.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    SHARE           -> image "./assets/images/ControllerIcons/PlaystationButtons/Share.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    HOME_BUTTON     -> image "./assets/images/ControllerIcons/PlaystationButtons/PSButton.png"    `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    TRACKPAD        -> image "./assets/images/ControllerIcons/PlaystationButtons/Touchbar.png"    `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50] --Playstation Only
    PS_BUTTON_A     -> image "./assets/images/ControllerIcons/PlaystationButtons/X.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_B     -> image "./assets/images/ControllerIcons/PlaystationButtons/Circle.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_X     -> image "./assets/images/ControllerIcons/PlaystationButtons/Square.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_Y     -> image "./assets/images/ControllerIcons/PlaystationButtons/Triangle.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_L1    -> image "./assets/images/ControllerIcons/PlaystationButtons/L1.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_R1    -> image "./assets/images/ControllerIcons/PlaystationButtons/R1.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_L2    -> image "./assets/images/ControllerIcons/PlaystationButtons/L2.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_R2    -> image "./assets/images/ControllerIcons/PlaystationButtons/R2.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_L3    -> image "./assets/images/ControllerIcons/PlaystationButtons/L3.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    PS_BUTTON_R3    -> image "./assets/images/ControllerIcons/PlaystationButtons/R3.png"      `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]

thirdPartyList c =   
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]        
    KEYBOARD_KEY    -> image "./assets/images/ControllerIcons/keyboard_key.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    L_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_left_click.png"   `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    R_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_right_click.png"  `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    C_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_middle_click.png" `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    XR_GESTURE      -> image "./assets/images/ControllerIcons/xr_gesture.png"         `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]  

genericKeyList c =   
  case c of
    Mapping         -> image "./assets/images/ControllerIcons/mapping_icon.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]        
    KEYBOARD_KEY    -> image "./assets/images/ControllerIcons/keyboard_key.png"       `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    L_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_left_click.png"   `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    R_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_right_click.png"  `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    C_MOUSE_KEY     -> image "./assets/images/ControllerIcons/mouse_middle_click.png" `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]
    XR_GESTURE      -> image "./assets/images/ControllerIcons/xr_gesture.png"         `styleBasic` [sizeReqW $ fixedSize 50, sizeReqH $ fixedSize 50]  

appModel :: BooksModel
appModel   = initialModel

doneBg :: Color
doneBg     = rgbHex "#CFF6E2"

doneFg :: Color
doneFg     = rgbHex "#459562"

pendingBg :: Color
pendingBg  = rgbHex "#F5F0CC"

pendingFg :: Color
pendingFg  = rgbHex "#827330"

grayLight :: Color
grayLight  = rgbHex "#9E9E9E"

grayDark  :: Color
grayDark   = rgbHex "#393939"

grayDarker :: Color
grayDarker = rgbHex "#2E2E2E"

customLightTheme :: Theme
customLightTheme = lightTheme
  & L.userColorMap . at "rowBgColor" ?~ rgbHex "#ECECEC"

customDarkTheme :: Theme
customDarkTheme = darkTheme
  & L.userColorMap . at "rowBgColor" ?~ rgbHex "#656565"
  