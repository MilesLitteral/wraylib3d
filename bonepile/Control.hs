{-# LANGUAGE AllowAmbiguousTypes #-}

module Manifest.Control where

import Manifest.Control.NetworkManager
import Manifest.Control.UDP
