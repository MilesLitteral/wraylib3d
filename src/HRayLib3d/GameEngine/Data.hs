module HRayLib3d.GameEngine.Data (
        -- module HRayLib3d.GameEngine.Data.BSP,
        -- module HRayLib3d.GameEngine.Data.GameCharacter,
        -- module HRayLib3d.GameEngine.Data.GLB,
        -- module HRayLib3d.GameEngine.Data.GLTF,
        -- module HRayLib3d.GameEngine.Data.Material,
        -- module HRayLib3d.GameEngine.Data.MD3,
        -- module HRayLib3d.GameEngine.Data.OBJ
    ) where

    -- import HRayLib3d.GameEngine.Data.BSP
    -- import HRayLib3d.GameEngine.Data.GameCharacter
    -- import HRayLib3d.GameEngine.Data.GLB
    -- import HRayLib3d.GameEngine.Data.GLTF
    -- import HRayLib3d.GameEngine.Data.Material
    -- import HRayLib3d.GameEngine.Data.MD3
    -- import HRayLib3d.GameEngine.Data.OBJ