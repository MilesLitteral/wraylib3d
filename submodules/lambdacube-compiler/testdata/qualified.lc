{-# LANGUAGE NoImplicitPrelude #-}

import qualified Prelude as P
import qualified Prelude hiding (a,b) as PH
import qualified Prelude (c,e)
-- import qualified Prelude (c,e) hiding (a,b) as PH -- fails
