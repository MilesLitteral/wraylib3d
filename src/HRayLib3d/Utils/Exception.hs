module HRayLib3d.Utils.Exception (try', catchAny) where

    import Control.Exception

    -- Misc. Exception catching code:

    --try' can be used to catch an exception in any IO function
    try' :: IO a ->  IO (Either IOException a)
    try' =  try 

    -- Avoid the "Ambiguous type variable..." error
    catchAny :: IO a -> (SomeException -> IO a) -> IO a
    catchAny = catch