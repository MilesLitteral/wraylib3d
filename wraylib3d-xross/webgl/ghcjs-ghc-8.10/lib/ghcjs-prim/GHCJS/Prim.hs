{-# LANGUAGE MagicHash, DeriveDataTypeable, ScopedTypeVariables, CPP #-}
#ifdef __GHCJS__
{-# LANGUAGE JavaScriptFFI,
             GHCForeignImportPrim,
             UnliftedFFITypes,
             UnboxedTuples
  #-}
#endif

module GHCJS.Prim ( JSVal(..), JSVal#
                  , JSException(..)
                  , WouldBlockException(..)
#ifdef ghcjs_HOST_OS
                  , toIO
                  , resolve
                  , resolveIO
                  , mkJSException
                  , fromJSString
                  , toJSString
                  , toJSArray
                  , fromJSArray
                  , fromJSInt
                  , toJSInt
                  , isNull
                  , isUndefined
                  , jsNull
                  , getProp
                  , getProp'
                  , getProp#
                  , unsafeGetProp
                  , unsafeGetProp'
                  , unsafeGetProp#
                  , unpackJSString#
                  , unpackJSStringUtf8#
                  , unsafeUnpackJSString#
                  , unsafeUnpackJSStringUtf8#
                  , unpackJSStringUtf8##
                  , unsafeUnpackJSStringUtf8##

#endif
                  ) where

import           Data.Typeable (Typeable)
import           Unsafe.Coerce (unsafeCoerce)

import           GHC.Prim
import qualified GHC.Exception as Ex
import qualified GHC.Exts as Exts
import qualified GHC.CString as GHC
import           GHC.IO

{-
  JSVal is a boxed type that can be used as FFI
  argument or result.
-}
#ifdef ghcjs_HOST_OS
data JSVal  = JSVal ByteArray#
type JSVal# = ByteArray#
#else
data JSVal  = JSVal Addr#
type JSVal# = Addr#
#endif

{-
  When a JavaScript exception is raised inside
  a safe or interruptible foreign call, it is converted
  to a JSException
 -}
data JSException = JSException JSVal String
  deriving (Typeable)

instance Ex.Exception JSException

instance Show JSException where
  show (JSException _ xs) = "JavaScript exception: " ++ xs

#ifdef ghcjs_HOST_OS

{-# NOINLINE toIO #-}
toIO :: Exts.Any -> IO Exts.Any
toIO x = pure x

{-# NOINLINE resolve #-}
resolve :: JSVal# -> JSVal# -> Exts.Any -> IO ()
resolve accept reject x = resolveIO accept reject (pure x)

{-# NOINLINE resolveIO #-}
resolveIO :: JSVal# -> JSVal# -> IO Exts.Any -> IO ()
resolveIO accept reject x =
  (x >>= evaluate >>= js_callback_any accept) `catch`
  (\(e::Ex.SomeException) -> do
    exceptionText <- evaluate (toJSString $ Ex.displayException e) `catch`
                         (\(_::Ex.SomeException) -> evaluate (toJSString "unknown exception"))
    js_callback_jsval reject exceptionText)

mkJSException :: JSVal -> IO JSException
mkJSException ref =
  return (JSException (unsafeCoerce ref) (fromJSString ref))

{- | Low-level conversion utilities for packages that cannot
     depend on ghcjs-base
 -}

{- | returns an empty string if the JSVal does not contain
     a string
 -}
fromJSString :: JSVal -> String
fromJSString = unsafeCoerce . js_fromJSString
{-# INLINE fromJSString #-}

toJSString :: String -> JSVal
toJSString = js_toJSString . unsafeCoerce . seqList
{-# INLINE [0] toJSString #-}
{-# RULES
"GHCJSPRIM toJSString/literal" forall a.
  toJSString (GHC.unpackCString# a) = JSVal (unsafeUnpackJSStringUtf8## a)
"GHCJSPRIM toJSString/literalUtf8" forall a.
  toJSString (GHC.unpackCStringUtf8# a) = JSVal (unsafeUnpackJSStringUtf8## a)
  #-}

fromJSArray :: JSVal -> IO [JSVal]
fromJSArray = unsafeCoerce . js_fromJSArray
{-# INLINE fromJSArray #-}

toJSArray :: [JSVal] -> IO JSVal
toJSArray = js_toJSArray . unsafeCoerce . seqList
{-# INLINE toJSArray #-}

{- | returns zero if the JSVal does not contain a number
 -}
fromJSInt :: JSVal -> Int
fromJSInt = js_fromJSInt
{-# INLINE fromJSInt #-}

toJSInt :: Int -> JSVal
toJSInt = js_toJSInt
{-# INLINE toJSInt #-}

isNull :: JSVal -> Bool
isNull = js_isNull
{-# INLINE isNull #-}

isUndefined :: JSVal -> Bool
isUndefined = js_isUndefined
{-# INLINE isUndefined #-}

jsNull :: JSVal
jsNull = js_null
{-# INLINE CONLIKE jsNull #-}

getProp :: JSVal -> String -> IO JSVal
getProp o p = js_getProp o (unsafeCoerce $ seqList p)
{-# INLINE [0] getProp #-}
{-# RULES
"GHCJSPRIM getProp/literal" forall o a.
  getProp o (GHC.unpackCString# a) = getProp# o a
"GHCJSPRIM getProp/literalUtf8" forall o a.
  getProp o (GHC.unpackCStringUtf8# a) = getPropUtf8# o a
  #-}

-- | only safe on immutable object
unsafeGetProp :: JSVal -> String -> JSVal
unsafeGetProp o p = js_unsafeGetProp o (unsafeCoerce $ seqList p)
{-# INLINE [0] unsafeGetProp #-}
{-# RULES
"GHCJSPRIM unsafeGetProp/literal" forall o a.
  unsafeGetProp o (GHC.unpackCString# a) = unsafeGetProp# o a
"GHCJSPRIM unsafeGetProp/literalUtf8" forall o a.
  unsafeGetProp o (GHC.unpackCStringUtf8# a) = unsafeGetPropUtf8# o a
  #-}

getProp' :: JSVal -> JSVal -> IO JSVal
getProp' o p = js_getProp' o p
{-# INLINE [0] getProp' #-}
{-# RULES
"GHCJSPRIM getProp'/literal" forall o a.
  getProp' o (unsafeUnpackJSString# a) = getProp# o a
"GHCJSPRIM getProp'/literalUtf8" forall o a.
  getProp' o (unsafeUnpackJSStringUtf8# a) = getPropUtf8# o a
  #-}

-- | only safe on immutable object
unsafeGetProp' :: JSVal -> JSVal -> JSVal
unsafeGetProp' o p = js_unsafeGetProp' o p
{-# INLINE [0] unsafeGetProp' #-}
{-# RULES
"GHCJSPRIM unsafeGetProp'/literal" forall o a.
  unsafeGetProp' o (unsafeUnpackJSString# a) = unsafeGetPropUtf8# o a
"GHCJSPRIM unsafeGetProp'/literalUtf8" forall o a.
  unsafeGetProp' o (unsafeUnpackJSStringUtf8# a) = unsafeGetPropUtf8# o a
  #-}


-- | only safe on immutable Addr#
getProp# :: JSVal -> Addr# -> IO JSVal
getProp# (JSVal o) p = IO $
  \s -> case getPropUtf8## o p s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] getProp# #-}
-- js_getProp# o p

-- | only safe on immutable Addr#
getPropUtf8# :: JSVal -> Addr# -> IO JSVal
getPropUtf8# (JSVal o) p = IO $
  \s -> case getPropUtf8## o p s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] getPropUtf8# #-}

getPropUtf8## :: JSVal# -> Addr# -> State# s -> (# State# s, JSVal# #)
getPropUtf8## o p = js_getPropUtf8## o p
{-# NOINLINE getPropUtf8## #-}

-- | only safe on immutable Addr# and JSVal
unsafeGetProp# :: JSVal -> Addr# -> JSVal
unsafeGetProp# (JSVal o) p = JSVal (unsafeGetPropUtf8## o p)
{-# INLINE [0] unsafeGetProp# #-}

-- | only safe on immutable Addr# and JSVal
unsafeGetPropUtf8# :: JSVal -> Addr# -> JSVal
unsafeGetPropUtf8# (JSVal o) p = JSVal (unsafeGetPropUtf8## o p)
{-# INLINE [0] unsafeGetPropUtf8# #-}

unsafeGetPropUtf8## :: JSVal# -> Addr# -> JSVal#
unsafeGetPropUtf8## o p = js_unsafeGetPropUtf8## o p
{-# NOINLINE unsafeGetPropUtf8## #-}

unpackJSString# :: Addr# -> IO JSVal
unpackJSString# a = IO $
  \s -> case unpackJSStringUtf8## a s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] unpackJSString# #-}

unpackJSStringUtf8# :: Addr# -> IO JSVal
unpackJSStringUtf8# a = IO $
  \s -> case unpackJSStringUtf8## a s of (# s', v #) -> (# s', JSVal v #)
{-# INLINE [0] unpackJSStringUtf8# #-}

unpackJSStringUtf8## :: Addr# -> State# s -> (# State# s, JSVal# #)
unpackJSStringUtf8## a s = js_unpackJSStringUtf8## a s
{-# NOINLINE unpackJSStringUtf8## #-}

-- | only safe on immutable Addr#
unsafeUnpackJSString# :: Addr# -> JSVal
unsafeUnpackJSString# a = JSVal (unsafeUnpackJSStringUtf8## a)
  -- js_unsafeUnpackJSString# a
{-# INLINE [0] unsafeUnpackJSString# #-}

-- | only safe on immutable Addr#
unsafeUnpackJSStringUtf8# :: Addr# -> JSVal
unsafeUnpackJSStringUtf8# a = JSVal (unsafeUnpackJSStringUtf8## a)
{-# INLINE [0] unsafeUnpackJSStringUtf8# #-}

unsafeUnpackJSStringUtf8## :: Addr# -> JSVal#
unsafeUnpackJSStringUtf8## a = js_unsafeUnpackJSStringUtf8## a
{-# NOINLINE unsafeUnpackJSStringUtf8## #-}


-- reduce the spine and all list elements to whnf
seqList :: [a] -> [a]
seqList xs = go xs `seq` xs
  where go (x:xs) = x `seq` go xs
        go []     = ()

seqListSpine :: [a] -> [a]
seqListSpine xs = go xs `seq` xs
  where go (x:xs) = go xs
        go []     = ()

foreign import javascript unsafe "h$toHsString($1)"
  js_fromJSString :: JSVal -> Exts.Any

foreign import javascript unsafe "h$fromHsString($1)"
  js_toJSString :: Exts.Any -> JSVal

foreign import javascript unsafe "h$toHsListJSVal($1)"
  js_fromJSArray :: JSVal -> IO Exts.Any

foreign import javascript unsafe "h$fromHsListJSVal($1)"
  js_toJSArray :: Exts.Any -> IO JSVal

foreign import javascript unsafe "$1 === null"
  js_isNull :: JSVal -> Bool

foreign import javascript unsafe "$1 === undefined"
  js_isUndefined :: JSVal -> Bool

foreign import javascript unsafe "$r = typeof($1) === 'number' ? ($1|0) : 0;"
  js_fromJSInt :: JSVal -> Int

foreign import javascript unsafe "$r = $1;"
  js_toJSInt :: Int -> JSVal

foreign import javascript unsafe "$r = null;"
  js_null :: JSVal

foreign import javascript unsafe "$1[h$fromHsString($2)]"
  js_getProp :: JSVal -> Exts.Any -> IO JSVal

foreign import javascript unsafe "$1[h$fromHsString($2)]"
  js_unsafeGetProp :: JSVal -> Exts.Any -> JSVal

foreign import javascript unsafe "$1[$2]"
  js_getProp' :: JSVal -> JSVal -> IO JSVal

foreign import javascript unsafe "$1[$2]"
  js_unsafeGetProp' :: JSVal -> JSVal -> JSVal

foreign import javascript unsafe "$1[h$decodeUtf8z($2_1, $2_2)]"
  js_getPropUtf8## :: JSVal# -> Addr# -> State# s -> (# State# s, JSVal# #)

foreign import javascript unsafe "$1[h$decodeUtf8z($2_1, $2_2)]"
  js_unsafeGetPropUtf8## :: JSVal# -> Addr# -> JSVal#

foreign import javascript unsafe "h$decodeUtf8z($1_1, $1_2)"
  js_unpackJSStringUtf8## :: Addr# -> State# s -> (# State# s, JSVal# #)


foreign import javascript unsafe "h$decodeUtf8z($1_1, $1_2)"
  js_unsafeUnpackJSStringUtf8## :: Addr# -> JSVal#

foreign import javascript unsafe "$1($2)"
  js_callback_any :: JSVal# -> Exts.Any -> IO ()

foreign import javascript unsafe "$1($2)"
  js_callback_jsval :: JSVal# -> JSVal -> IO ()

#endif

{- | If a synchronous thread tries to do something that can only
     be done asynchronously, and the thread is set up to not
     continue asynchronously, it receives this exception.
 -}
data WouldBlockException = WouldBlockException
  deriving (Typeable)

instance Show WouldBlockException where
  show _ = "thread would block"

instance Ex.Exception WouldBlockException
