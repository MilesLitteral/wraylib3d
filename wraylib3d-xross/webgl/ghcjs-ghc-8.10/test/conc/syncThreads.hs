{-# LANGUAGE ScopedTypeVariables, JavaScriptFFI, ForeignFunctionInterface #-}

module Main where

import Prelude hiding (print, putStrLn)

import Control.Concurrent
import Control.Concurrent.MVar
import Control.Exception
import Control.Monad

import Data.List (intersperse)

import GHCJS.Types
import GHCJS.Concurrent
import GHCJS.Foreign.Callback

import qualified Data.JSString as JSS

---------------------------------------------------------------------------
-- our usual standard io implementation is asynchronous, use this more
-- primitive mechanism to print the results

print :: Show a => a -> IO ()
print = putStrLn . show

putStrLn :: String -> IO ()
putStrLn = js_log . JSS.pack

foreign import javascript unsafe
  "console.log($1);"
  js_log :: JSString -> IO ()

---------------------------------------------------------------------------

foreign import javascript unsafe
  "$1();"
  js_runCallback :: Callback a -> IO ()

---------------------------------------------------------------------------

main :: IO ()
main = sequence_ .
       intersperse (putStrLn "---------------------------") $
         [ synchronously1
         , wouldBlock1
         , wouldBlock2
         , callback1
         , callback2
         , callback3
         ]

printIsSync :: IO ()
printIsSync = do
  tid  <- myThreadId
  sync <- isThreadSynchronous tid
  ca   <- isThreadContinueAsync tid
  print (sync, ca)

printMVar :: MVar () -> IO ()
printMVar = print <=< takeMVar

synchronously1 :: IO ()
synchronously1 = do
  putStrLn "synchronously1"
  printIsSync
  synchronously printIsSync
  printIsSync
  let h x m = m `catch` \(e::SomeException) -> do
                           putStrLn ("handler: " ++ x)
                           printIsSync
  h "outside" (synchronously $ printIsSync >> error "err")
  printIsSync
  synchronously (h "inside" (printIsSync >> error "err"))
  printIsSync
  putStrLn "synchronously1 done"
  
-- blocking on MVar should give us an exception, the exception can be handled
wouldBlock1 :: IO ()
wouldBlock1 = do
  putStrLn "wouldBlock1"
  x `catch` \(e::SomeException) ->
    do putStrLn "exception caught: outer"
       print e
  putStrLn "wouldBlock1 done"
  where
    x = do
      mv1 <- newEmptyMVar
      printIsSync
      synchronously $ do
        printIsSync
        printMVar mv1 `catch` \(e::SomeException) -> do
          putStrLn "exeption caught inner1"
          print e
        printMVar mv1 `catch` \(e::WouldBlockException) -> do
          putStrLn "exeption caught inner2"
          print e
        putStrLn "ok"
        printMVar mv1
        putStrLn "unreachable"
      printIsSync

-- threadDelay should give us the exception too
wouldBlock2 :: IO ()
wouldBlock2 = do
  putStrLn "wouldBlock2"
  threadDelay 500000
  let x = synchronously $ do
            threadDelay 500000 `catch` \(e::WouldBlockException) ->
              putStrLn "exception caught: inner"
            printIsSync
            putStrLn "ok"
            threadDelay 500000
            putStrLn "unreachable"
  x `catch` \(e::WouldBlockException) ->
    putStrLn "exception caught: outer"
  putStrLn "wouldBlock2 done"

-- synchronous callbacks give us an exception
callback1 :: IO ()
callback1 = do
  putStrLn "callback1"
  mv1 <- newEmptyMVar
  cb1 <- syncCallback ThrowWouldBlock $ do
    printIsSync
    putStrLn "ok"
    printMVar mv1 `catch` \(e::WouldBlockException) ->
      putStrLn "exception: would block"
    printIsSync
    putStrLn "ok"
    printMVar mv1
    -- thread would block error ends up on stderr
    putStrLn "unreachable"
  js_runCallback cb1
  putStrLn "callback1 finished"
  releaseCallback cb1

callback2 :: IO ()
callback2 = do
  putStrLn "callback2"
  mv1 <- newEmptyMVar
  mv2 <- newEmptyMVar
  cb1 <- syncCallback ContinueAsync $ do
    printIsSync
    putStrLn "ok"
    printMVar mv1
    putStrLn "callback"
    printIsSync
    putMVar mv2 ()
  js_runCallback cb1
  putStrLn "main"
  putMVar mv1 ()
  printMVar mv2
  putStrLn "main"
  putStrLn "callback2 done"  

-- async callbacks are async
callback3 :: IO ()
callback3 = do
  putStrLn "callback3"
  mv1 <- newEmptyMVar
  cb1 <- asyncCallback $ do
    putStrLn "async callback"
    printIsSync
    putMVar mv1 ()
    printMVar mv1
    putStrLn "callback"
    putMVar mv1 ()
  js_runCallback cb1
  printMVar mv1
  putStrLn "main"
  putMVar mv1 ()
  printMVar mv1
  putStrLn "main"
  printIsSync
  releaseCallback cb1
  putStrLn "callback3 done"
