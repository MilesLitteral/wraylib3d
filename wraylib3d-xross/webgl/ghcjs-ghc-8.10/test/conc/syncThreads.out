synchronously1
(False,True)
(True,False)
(False,True)
(True,False)
handler: outside
(False,True)
(False,True)
(True,False)
handler: inside
(True,False)
(False,True)
synchronously1 done
---------------------------
wouldBlock1
(False,True)
(True,False)
exeption caught inner1
thread would block
exeption caught inner2
thread would block
ok
exception caught: outer
thread would block
wouldBlock1 done
---------------------------
wouldBlock2
exception caught: inner
(True,False)
ok
exception caught: outer
wouldBlock2 done
---------------------------
callback1
(True,False)
ok
exception: would block
(True,False)
ok
callback1 finished
---------------------------
callback2
(True,True)
ok
main
()
callback
(False,True)
()
main
callback2 done
---------------------------
callback3
async callback
(False,True)
()
main
()
callback
()
main
(False,True)
callback3 done
